zlib: add zstd support

jkrems opened the pull request:
Adds `ZstdCompress` and `ZsdDecompress` to the `zlib` module which can be used to compress/decompress with the Zstandard ("zstd") algorithm.

Notable omissions:

* Providing dictionaries isn't implemented.
* The docs in `zlib.md` don't call out any params beyond the basic compression level.

The code follows similar patterns to the [PR that added Brotli support](https://github.com/nodejs/node/pull/24938). Just that instead of brotli, it adds the equivalent zstd APIs. Just like Brotli, this required separate compression/decompression context objects.

~~Zstd itself has been around and stable for multiple years but this PR is early in terms of web support: It only just starts shipping by default in Chrome 123. On the other hand, by shipping in Chrome it will soon be supported quite widely on the web. Firefox also signaled support (https://bugzilla.mozilla.org/show_bug.cgi?id=1301878#c65).~~ Zstd is now enabled on the web and supported by both Chrome and Firefox: https://caniuse.com/zstd

Official support in node.js would allow passing additional WPTs around `fetch` (https://github.com/nodejs/undici/issues/2847).

<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

jkrems commented:
Alright, got up to making the basics (seemingly) work:

```console
$ out/Debug/node -p 'zlib.zstdDecompressSync(zlib.zstdCompressSync("Hello World")).toString()'
Hello World
```

terrelln commented on src/node_zlib.cc:
```diff
@@ -1253,6 +1434,109 @@ CompressionError BrotliDecoderContext::GetErrorInfo() const {
+  }
+}
+
+CompressionError ZstdCompressContext::SetParameter(int key, uint32_t value) {
```
`ZSTD_CCtx_setParameter()` takes an `int` value. Some functions do want to allow negative values, like the compression level set via `ZSTD_c_compressionLevel`.

jkrems replied:
Thanks, updated to use `int` here.

terrelln commented on src/node_zlib.cc:
```diff
@@ -1253,6 +1434,109 @@ CompressionError BrotliDecoderContext::GetErrorInfo() const {
+  }
+}
+
+CompressionError ZstdDecompressContext::SetParameter(int key, uint32_t value) {
```
Same here, the value for `ZSTD_DCtx_setParameter()` is an `int`.

jkrems replied:
Thanks, updated to use `int` here.

terrelln reviewed (commented):
I've only looked at the usage of the Zstd API. It looks about right, I've left some minor comments. I'm not familiar with the code, and it isn't clear to me exactly how `DoThreadPoolWork()` is called, so I can't be completely confident. But broadly it looks right.

One thing to consider is that some advanced parameters directly can cause Zstd to allocate a lot of memory (e.g. setting `ZSTD_c_hashLog == ZSTD_c_chainLog == ZSTD_c_windowLog == 31` will cause Zstd to allocate at least 6 GB during compression). Additionally, the parameter `ZSTD_c_nbWorkers` will cause Zstd to spawn worker threads during compression.

This normally isn't an issue, because the author of the code is trusted. I assume that this is also the threat model for NodeJS. But, because it is JS, I did want to raise the point.

terrelln commented:
> There's no .params() support even though zstd should support changing (some) params at runtime.

There must be at least compression level support `ZSTD_c_compressionLevel`. This is the most important parameter by far.

What do you mean at runtime? You can change any parameters you want before the (de)compression starts. After that, changing compression parameters generally isn't allowed.

> No input size hints for compression because it looked like that API in zstd wasn't stable yet.

If the input size is known in advance, then you can use [`ZSTD_setPledgedSrcSize()`](https://github.com/facebook/zstd/blob/6a0052a409e2604bd40354b76b86272b712edd7d/lib/zstd.h#L546-L561), however it is an error to pass an incorrect value. Additionally, if the first call to `ZSTD_compressStream2()` uses `ZSTD_e_end`, Zstd will infer the input size.

If the input size is only approximately known, but may be wrong, then Zstd also supports setting `ZSTD_c_srcSizeHint`. Zstd will use this parameter for tuning its tradeoffs for that source size, but it is allowed to be wrong. But, as you say this is a newer feature that we haven't yet stabilized. It works and is fully supported, but all new features go through our experimental API before being stabilized, so we have the freedom to tweak or remove them in the future.

jkrems commented:
> What do you mean at runtime? You can change any parameters you want before the (de)compression starts. After that, changing compression parameters generally isn't allowed.

Thanks for confirming. The "at runtime" refers to the deflate-specific feature (`deflateParams` in zlib) that allows changing certain params within the same compression stream (at block boundaries IIUC). Node-level docs: https://nodejs.org/api/zlib.html#zlibparamslevel-strategy-callback

Updated the PR description to correct this part.

> If the input size is known in advance, then you can use ZSTD_setPledgedSrcSize(), however it is an error to pass an incorrect value. Additionally, if the first call to ZSTD_compressStream2() uses ZSTD_e_end, Zstd will infer the input size.

Ah, thanks. Saw `ZSTD_c_srcSizeHint` but should've kept looking. Added `pledgedSrcSize` as an option.

jkrems commented:
Rebased after the fixes to ensure it still lands cleanly.

jkrems commented:
Looking at the test failures, they look unrelated at first glance (different tests on different OS, not related to compression afaict). Resuming the build to see if the are flakes.

bricss commented:
Maybe it's possible to bump `zstd` to the latest version? 🤔

jkrems commented:
Happy to rerun the import. I think right now I'm also looking for signal that this is something that other collaborators want in core. Otherwise rebasing (potentially) and reimporting wouldn't really be worth it - it would just bitrot again.

bricss commented:
I cannot speak on behalf of collaborators, but imho W3C community would certainly like to see it in the core 🧐
Given the fact that [Zstd Content-Encoding](https://chromestatus.com/feature/6186023867908096) will soon be [supported](https://caniuse.com/zstd) by majority of modern browsers 🧭

KhafraDev commented:
cc @nodejs/zlib @nodejs/security-wg

anonrig commented on src/node_zlib.cc:
```diff
@@ -1253,6 +1454,117 @@ CompressionError BrotliDecoderContext::GetErrorInfo() const {
+
+CompressionError ZstdCompressContext::Init(uint64_t pledged_src_size) {
+  pledged_src_size_ = pledged_src_size;
+  cctx_.reset(ZSTD_createCCtx());
```
Why cctx?

```suggestion
  ctx.reset(ZSTD_createCCtx());
```

jkrems replied:
The `c` is for compression (cctx vs dctx). Given that we only ever deal with one or the other, we could drop that marker from the fields.

anonrig commented on src/node_zlib.cc:
```diff
@@ -1253,6 +1454,117 @@ CompressionError BrotliDecoderContext::GetErrorInfo() const {
+    return CompressionError(
+        "Could not set pledged src size", "ERR_ZLIB_INITIALIZATION_FAILED", -1);
+  }
+  return CompressionError{};
```
```suggestion
  return {};
```

anonrig commented on src/node_zlib.cc:
```diff
@@ -1253,6 +1454,117 @@ CompressionError BrotliDecoderContext::GetErrorInfo() const {
+    error_ = ZSTD_getErrorCode(remaining);
+    error_code_string_ = ZstdStrerror(error_);
+    error_string_ = ZSTD_getErrorString(error_);
+    return;
```
```suggestion
```

anonrig commented on src/node_zlib.cc:
```diff
@@ -1253,6 +1454,117 @@ CompressionError BrotliDecoderContext::GetErrorInfo() const {
+    return CompressionError(
+        "Setting parameter failed", "ERR_ZSTD_PARAM_SET_FAILED", -1);
+  }
+  return CompressionError{};
```
ditto

anonrig commented on src/node_zlib.cc:
```diff
@@ -1253,6 +1454,117 @@ CompressionError BrotliDecoderContext::GetErrorInfo() const {
+                            "ERR_ZLIB_INITIALIZATION_FAILED",
+                            -1);
+  }
+  return CompressionError{};
```
ditto

anonrig commented on src/node_zlib.cc:
```diff
@@ -1253,6 +1454,117 @@ CompressionError BrotliDecoderContext::GetErrorInfo() const {
+    error_ = ZSTD_getErrorCode(ret);
+    error_code_string_ = ZstdStrerror(error_);
+    error_string_ = ZSTD_getErrorString(error_);
+    return;
```
```suggestion
```

anonrig commented on src/node_zlib.cc:
```diff
@@ -248,6 +290,79 @@ class BrotliDecoderContext final : public BrotliContext {
+  CompressionError SetParameter(int key, int value);
+
+  // Wrap ZSTD_freeCCtx to remove the return type.
+  static void FreeCCtx(ZSTD_CCtx* cctx) { ZSTD_freeCCtx(cctx); }
```
```suggestion
  static void FreeZstd(ZSTD_CCtx* cctx) { ZSTD_freeCCtx(cctx); }
```

anonrig commented:
Can you also update license_builder?

jkrems commented:
> Can you also update license_builder?

The changes to the top-level LICENSE file were generated by running `./tools/license-builder.sh`. It's at the very end of the file changes in the Github UI.

jkrems commented:
Moved the init errors into C++ where we can add more context to them.

maxpain commented:
Hello. Any ETA for this?

jasnell commented:
@jkrems ... can you give this a rebase?

jkrems commented:
@jasnell I can but at this point I have spend quite bit of time on this PR without a clear signal on whether this would be likely to land (in this form or another). So it would be easier to justify if there was a tentative "yep, the high level approach looks good". By high level I mean something like "roughly following brotli integration in terms of API, using the facebook zstandard native implementation as the backing".

bricss commented:
That's will be great to have this in a `main` branch 🌿after all 🫠

KhafraDev commented:
@nodejs/tsc There aren't any outstanding objections, but this PR does appear to be at an impasse. @jkrems did not receive an answer to their question https://github.com/nodejs/node/pull/52100#issuecomment-2458432923. I have held off for months from doing this, hoping someone could review and collaborate since I couldn't.

marco-ippolito commented:
Nobody reviewed and approved the PR, there is nothing officially blocking it. If someone wants to volunteer to review it, then its gonna land

KhafraDev commented:
Maybe not, but I don't think I'm in a position to give guidance or give definitive instructions on the approach. I have pinged the zlib team, I pinged the security team, I have asked people to review it in private, so I didn't know what else to do lol.

anonrig commented:
I'm interested in landing this PR as well. If we can have one last push (and rebase this PR), I'm more than happy to review and land this PR.

jkrems commented:
@anonrig Thanks! I'll work on a rebase.
